### ISSUE TITLE
Coordinates for labels clarification 

### ISSUE THREAD

**Issue by yzhou377** [argoai/argoverse-api#19]
Greetings. 

As I am working with the data, I could not find the coordinates for which labels are used. I noticed that you did a transformation to get them into the ego-vehicle frame, so I assume the labels are in LIDAR frame? Which LIDAR?

**Comment by johnwlambert** [comment]
Hi @yzhou377 , the cuboid labels are annotated in an object frame, and using the SE(3) transformation provided with the cuboid label, they can be moved from the object frame to the egovehicle frame. A good example is [here](https://github.com/argoai/argoverse-api/blob/master/demo_usage/cuboids_to_bboxes.py#L174), which uses the `ObjectLabelRecord` class. The `ObjectLabelRecord` class accepts label parameters, and can transform them into cuboid vertices living in the egovehicle frame (please see [ObjectLabelRecord.as_3d_bbox()](https://github.com/argoai/argoverse-api/blob/master/argoverse/data_loading/object_label_record.py#L78)

All LiDAR points are already living in the egovehicle frame, meaning their origin is the center of the back axle of the vehicle. You will not need to use the LiDAR SE(3) transformation matrices to work with the data, but we left them in the data for sake of completeness if people will to analyze or exploit parallax. 

For an example, please see [this file](https://github.com/argoai/argoverse-api/blob/master/demo_usage/visualize_ground_lidar_points.py)

If you load the LiDAR points, e.g. `lidar_pts = load_ply(ply_fpath)`, then since they are in the egovehicle frame ,you can call `project_lidar_to_undistorted_img(lidar_pts, ...)` [link](https://github.com/argoai/argoverse-api/blob/master/argoverse/utils/calibration.py#L498)
That function uses K [R | t ] to move LiDAR points into the camera coordinate system using `cam_SE3_egovehicle`, e.g. `uv_cam = cam_SE3_egovehicle.transform_point_cloud(points_egovehicle)`, and then finally moves them into image space, using `proj_cam_to_uv(uv_cam, camera_config)`.

We'll clarify this point -- thanks for the note.

**Comment by yzhou377** [comment]
Thank you for your quick reply John! BTW, I am glad to use one of our own GT data set, Go Jackets!

**Comment by yzhou377** [comment]
Hi John, I was working on a perception project a few days ago and asked you that question. I just finished an adapter to change the argoverse dataset format into KITTI dataset for more users in their established network testing or whatever. I have my adapter shared on my github here: https://github.com/yzhou377/argoverse-kitti-adapter Enjoy.

**Comment by johnwlambert** [comment]
Hi @yzhou377, thank you very much for your contribution.

Something we've noticed when playing with the data is that we often having an egovehicle driving past parked vehicles. The parked vehicles become occluded objects and are often objects with 3D vertices behind and in front of our camera sensors (positive and negative z). These introduce a few special cases to handle. I believe in your script you are only plotting the bounding boxes that are fully visible in the image (not the occluded ones). This would only render a few of the cuboids:

<img width="794" alt="Screen Shot 2019-07-02 at 12 23 20 AM" src="https://user-images.githubusercontent.com/16724970/60492344-a2488a80-9c5f-11e9-926f-dc5a18b8aa11.png">

A trick to get around this is to use our frustum clipping functions. I'll provide a code snippet below that will allow you to get those occluded bounding boxes, as follows:

<img width="591" alt="Screen Shot 2019-07-02 at 1 00 07 AM" src="https://user-images.githubusercontent.com/16724970/60494987-c5c20400-9c64-11e9-8a8f-268297060a6d.png">

Also I believe we haven't made our labeling policy very clear, I apologize for that. I noticed in your script that you were mentioning that 50 meters is *Maximum thresholding distance for labelled objects (Object beyond this distance will not be labelled)*. We actually do label objects far past 50 meters, as long as they lie within the region of interest (ROI), which we define as a 5 meter dilation of the driveable area isocontours (essentially within 5 meters of a curb).

**Comment by johnwlambert** [comment]
```python



import argoverse
from argoverse.data_loading.simple_track_dataloader import SimpleArgoverseTrackingDataLoader
from argoverse.data_loading.object_label_record import json_label_dict_to_obj_record
from argoverse.utils.calibration import CameraConfig
from argoverse.utils.transform import quat2rotmat
from argoverse.utils import calibration
from argoverse.data_loading.argoverse_tracking_loader import ArgoverseTrackingLoader

from argoverse.utils.calibration import (
    CameraConfig,
    get_calibration_config,
    point_cloud_to_homogeneous,
    project_lidar_to_img_motion_compensated,
    project_lidar_to_undistorted_img,
)
from argoverse.utils.frustum_clipping import (
	generate_frustum_planes,
	cuboid_to_2d_frustum_bbox
)
import copy
import imageio
import math
import matplotlib.pyplot as plt
import numpy as np
import os
from pathlib import Path
import pdb

RING_CAM_H = 1200
RING_CAM_W = 1920


def main(method = 'clipping'):
	""" """

	img_fpath = 'argoverse/argoverse-tracking/train1/53037376_5303_5303_5303_553038557184/ring_front_center/ring_front_center_315967823032038880.jpg'

	img = imageio.imread(img_fpath)

	log_id = '53037376_5303_5303_5303_553038557184'

	ply_fpath = 'argoverse/argoverse-tracking/train1/53037376_5303_5303_5303_553038557184/lidar/PC_315967823020014000.ply'
	cam_timestamp = 315967823032038880
	camera_name = 'ring_front_center'
	data_dir = 'argoverse/argoverse-tracking/train1'

	dl = SimpleArgoverseTrackingDataLoader(data_dir=data_dir, labels_dir=data_dir)
	calibration_data = dl.get_log_calibration_data(log_id)

	log_calib_data = dl.get_log_calibration_data(log_id)
	camera_config = get_calibration_config(log_calib_data, camera_name)
	K = camera_config.intrinsic
	planes = generate_frustum_planes(camera_config.intrinsic.copy(), camera_name)


	# load PLY file path, e.g. 'PC_315978406032859416.ply'
	ply_fpath = dl.get_closest_lidar_fpath(log_id, cam_timestamp)
	lidar_timestamp = Path(ply_fpath).stem.split("_")[-1]
	lidar_timestamp = int(lidar_timestamp)

	label_object_list = dl.get_labels_at_lidar_timestamp(log_id, lidar_timestamp)
	for i, label in enumerate(label_object_list):
		print(f'Plotting cuboid {i}')
		detected_object = json_label_dict_to_obj_record(label)
		classes = detected_object.label_class
		center = detected_object.translation # in ego frame 
		corners_ego_frame = detected_object.as_3d_bbox() # all eight points in ego frame 

		if method == 'fully_visible_only':
			calibration_data=calibration.load_calib(data_dir+'/'+log_id+'/vehicle_calibration_info.json')[camera_name]
			corners_cam_frame = calibration_data.project_ego_to_cam(corners_ego_frame) # all eight points in the camera frame 
			image_corners = calibration_data.project_ego_to_image(corners_ego_frame)
			image_bbox = [min(image_corners[:,0]), min(image_corners[:,1]),max(image_corners[:,0]),max(image_corners[:,1])]
			# the four coordinates we need for KITTI
			image_bbox = [round(x) for x in image_bbox]
			center_cam_frame = calibration_data.project_ego_to_cam(np.array([center]))
			in_bounds = 0<center_cam_frame[0][2] and 0<image_bbox[0]<RING_CAM_W and 0<image_bbox[1]<RING_CAM_H and 0<image_bbox[2]<RING_CAM_W and  0<image_bbox[3]<RING_CAM_H
			x1,y1,x2,y2 = round(image_bbox[0],2),round(image_bbox[1],2),round(image_bbox[2],2),round(image_bbox[3],2)
			if not in_bounds:
				continue
		
		elif method == 'clipping':

			points_h = point_cloud_to_homogeneous(corners_ego_frame).T
			_, uv_cam, _, _ = project_lidar_to_undistorted_img(points_h, copy.deepcopy(log_calib_data), camera_name)
			bbox_2d = cuboid_to_2d_frustum_bbox(uv_cam.T, planes, K[:3,:3])
			if bbox_2d is None:
				continue
			else:
				x1,y1,x2,y2 = bbox_2d

				x1 = min(x1,RING_CAM_W-1)
				x2 = min(x2,RING_CAM_W-1)
				y1 = min(y1,RING_CAM_H-1)
				y2 = min(y2,RING_CAM_H-1)

				x1 = max(x1, 0)
				x2 = max(x2, 0)
				y1 = max(y1, 0)
				y2 = max(y2, 0)

		plt.plot([x1,x1],[y1,y2], 'r')
		plt.plot([x1,x2],[y1,y1], 'r')
		plt.plot([x1,x2],[y2,y2], 'r')
		plt.plot([x2,x2],[y1,y2], 'r')

	plt.imshow(img)
	plt.savefig(f'all_now_clipping_green.png')
	plt.close('all')



	




if __name__ == '__main__':
	main()










```

**Comment by johnwlambert** [comment]
The image I rendered is rendered here: https://youtu.be/_A67O3hHVaY, about 10 seconds in.

**Comment by johnwlambert** [comment]
The reason we need the 3d frustum clipping is because if we just clip the `min_x, min_y, max_x, max_y` cuboids vertex coordinates to the image boundaries, if an object is shrinking in the image (like the bus on the left), we will end up with an oversized box (see left) vs. true box (see right) from 3d frustum clipping. Note that the cuboid annotation is a bit oversized.
<img width="1217" alt="Screen Shot 2019-07-02 at 9 29 28 AM" src="https://user-images.githubusercontent.com/16724970/60529732-1577ee00-9cac-11e9-9a88-b42f16b694e8.png">

**Comment by awethaileslassie** [comment]
@johnwlambert 
Hi, I am working on the Argoverse tracking dataset for 3D tasks such as Object detection and Tracking.
One thing I found and also expressed in the Argoverse dataset section is "Each sequence in our training and validation sets includes annotations for all objects within 5 meters of what we identify as the “drivable area” — the area in which it is possible for a vehicle to drive.", meaning even though my detection algorithm able to detect parked car > 5m away from the drivable area, it will be considered as FP, due to limitation of the corresponding annotation in the GT!!

I understudy that this can be masked using the maps, however, I am not utilizing the map, is there another alternative way to perform an evaluation on the annotated range?

**Comment by alliecc** [comment]
Hi @awethaileslassie,
Would you mind letting us know what is blocking you from using the map? All of our annotations come with LiDAR poses, which can be used to retrieve the map and mask out detections outside the ROI area.

**Comment by johnwlambert** [comment]
Hi @awethaileslassie, thanks for your interest in our work. The functionality you're requesting already exists in our API for detection evaluation:
https://github.com/argoai/argoverse-api/blob/master/argoverse/evaluation/detection/utils.py#L124

The ground truth and detected objects are filtered to the region of interest (ROI) before evaluation. Is there a specific reason you are not using our official evaluation scripts?
```python
    if cfg.eval_only_roi_instances and avm is not None:
        # go up 3 levels, because hierarchy is as follows:
        # {gt_root_fpath}/{log_id}/per_sweep_annotations_amodal/{gt_root_fname}
        gt_root_fpath = Path(gt_fpath).parents[2]
        city_SE3_egovehicle = get_city_SE3_egovehicle_at_sensor_t(ts, str(gt_root_fpath), log_id)
        if city_SE3_egovehicle is not None:
            log_city_name = read_city_name(os.path.join(gt_root_fpath, log_id, "city_info.json"))
            dts = filter_objs_to_roi(dts, avm, city_SE3_egovehicle, log_city_name)
            gts = filter_objs_to_roi(gts, avm, city_SE3_egovehicle, log_city_name)

```

**Comment by awethaileslassie** [comment]
Thank you for the feedback.

I was thinking to implement that during the training of the model so that it will only backpropagate for objects that have GT labels.

**Comment by johnwlambert** [comment]
@awethaileslassie I see, makes sense. You can use the logic we have in 
```python
filter_objs_to_roi(dts, avm, city_SE3_egovehicle, log_city_name)
```
to make a mask, and then elementwise multiply that with your loss values on each box, to prevent gradients from flowing to non-ROI dets.
